cpp/doc/libcuspatial_refactoring_guide.md

```diff
@@ -0,0 +1,272 @@
+OutputIt haversine_distance(LonLatItA a_lonlat_first,
+                            LonLatItA a_lonlat_last,
+                            LonLatItB b_lonlat_first,
+                            OutputIt distance_first,
+                            T const radius = EARTH_RADIUS_KM,
+                            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+```
+
+There are a few key points to notice.
+
+  1. The API is very similar to STL algorithms such as `std::transform`.
+  2. All array inputs and outputs are iterator type templates. 
+  3. Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
+     type (include/cuspatial/types.hpp)
+  4. The `Location` type is a template that is by default equal to the `value_type` of the input
+     iterators.
+  5. The floating point type is a template (`T`) that is by default equal to the `value_type` of
+     `Location`.
+  6. The iterator types for the two input ranges (A and B) are distinct templates. This is crucial
+     to enable composition of fancy iterators that may be different types for A and B.
+  7. The size of the input and output ranges in the example API are equal, so the start and end of
+     only the A range is provided (`a_lonlat_first` and `a_lonlat_last`). This mirrors STL APIs.
+  8. This API returns an iterator to the element past the last element written to the output. This
+     is inspired by `std::transform`, even though as with `transform`, many uses of 
+     `haversine_distance` will not need this returned iterator.
+  9. All APIs that run CUDA device code (including Thrust algorithms) or allocate memory take a CUDA
+     stream on which to execute the device code and allocate memory.
+
+## Example Documentation
+
+Following is the (Doxygen) documentation for the above `cuspatial::haversine_distance`.
+
+```C++
+/**
+ * @brief Compute haversine distances between points in set A to the corresponding points in set B.
+ *
+ * Computes N haversine distances, where N is `std::distance(a_lon_first, a_lon_last)`.
+ * The distance for each `(a_lon[i], a_lat[i])` and `(b_lon[i], b_lat[i])` point pair is assigned to
+ * `distance_first[i]`. `distance_first` must be an iterator to output storage allocated for N
+ * distances.
+ *
+ * https://en.wikipedia.org/wiki/Haversine_formula
+ *
+ * @param[in]  a_lonlat_first: beginning of range of (longitude, latitude) locations in set A
+ * @param[in]  a_lonlat_last: end of range of (longitude, latitude) locations in set A
+ * @param[in]  b_lonlat_first: beginning of range of (longitude, latitude) locations in set B
+ * @param[out] distance_first: beginning of output range of haversine distances
+ * @param[in]  radius: radius of the sphere on which the points reside. default: 6371.0
+ *            (approximate radius of Earth in km)
+ * @param[in]  stream: The CUDA stream on which to perform computations and allocate memory.
+ *
+ * All iterators must have the same floating-point `value_type`.
```

Comment:
This could be potentially confusing given that the `lonlat` iterators are expected to have a custom struct `value_type`.

Reply:
It's actually the `value_type` of the `value_type`. i.e you can't pass one iterator to `lonlat_2d<float>` and one to `lonlat_2d<double>`.

Reply:
Clarified.